src/dimension/mod.rs

```diff
@@ -689,6 +701,33 @@ where
     }
 }
 
+/// Attempt to merge axes if possible, starting from the back
+///
+/// Given axes [Axis(0), Axis(1), Axis(2), Axis(3)] this attempts
+/// to merge all axes one by one into Axis(3); when/if this fails,
+/// it attempts to merge the rest of the axes together into the next
+/// axis in line, for example a result could be:
+///
+/// [1, Axis(0) + Axis(1), 1, Axis(2) + Axis(3)] where `+` would
+/// mean axes were merged.
+pub(crate) fn merge_axes_from_the_back<D>(dim: &mut D, strides: &mut D)
+where
+    D: Dimension,
+{
+    debug_assert_eq!(dim.ndim(), strides.ndim());
+    match dim.ndim() {
+        0 | 1 => {}
+        n => {
+            let mut last = n - 1;
+            for i in (0..last).rev() {
+                if !merge_axes(dim, strides, Axis(i), Axis(last)) {
+                    last = i;
```

Comment:
I don't think that setting `last = i` will actually terminate the loop.

```suggestion
                    break;
```

Reply:
It just changes the axis we're currently merging into.

Reply:
Oh, I understand now. I was thinking that the intent was to merge into only the last axis.

Reply:
My bad, with the function name

Reply:
The function name is fine, IMO. It's more that I was thinking of the example code I wrote in [my comment on #932](https://github.com/rust-ndarray/ndarray/pull/932#discussion_r611237837), which did terminate the loop as soon as merging failed. I would suggest changing the variable name from `last` to `into` or `into_axis`, though.